SoundWire: start adding BPT/BRA support

bardliao opened the pull request:
This is a rebase version of https://github.com/thesofproject/linux/pull/4679
I removed the removal of ` select' statements at a higher level.` in the SND_SOC_SOF_HDA_LINK_BASELINE config in https://github.com/thesofproject/linux/pull/4679/commits/af779a3ca9bae844fb1691caaba7243e49afbe71 which I think it is removed by accident.

plbossart committed:
Documentation: driver: add SoundWire BRA description

The Bulk Register Access protocol was left as a TODO topic since
2018. It's time to document this protocol and the design of its Linux
support.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart committed:
soundwire: bus: add send_async/wait APIs for BPT protocol

Add definitions and helpers for the BPT/BRA protocol. Peripheral
drivers (aka ASoC codec drivers) can use this API to send bulk data
such as firmware or tables. The design intent is however NOT to
directly use this API but to rely on an intermediate regmap layer.

The API is only available when no other audio streams have been
allocated, and only one BTP/BRA stream is allowed per link. To avoid
the addition of yet another lock, the refcount tests are handled in
the stream master_runtime alloc/free routines where the bus_lock is
already held. Another benefit of this approach is that the same
bus_lock is used to handle runtime and port linked lists, which
reduces the potential for misaligned configurations.

In addition to exclusion with audio streams, BPT transfers have a lot
of overhead, specifically registers writes are needed to enable
transport in DP0. Most DMAs don't handle too well very small data sets
and they may have alignment limitations.

The size and alignment requirements are for now not handled by the
core but must be checked by platform-specific drivers.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart committed:
soundwire: debugfs: add interface for BPT/BRA transfers

Add code to show what codec drivers will need to do to enable BPT/BRA
transfers. The only difference is to set the 'command_type' file to
'1'. A zero-value will rely on regular read/write commands in Column0.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart committed:
ASoC: SOF: Intel: hda-sdw-bpt: add helpers for SoundWire BPT DMA

Add SoundWire BPT DMA helpers as a separate module to avoid circular
dependencies.

For now this assumes no link DMA, only coupled mode.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart committed:
soundwire: intel_ace2x: add BPT send_async/wait callbacks

Add support for BTP API using Cadence and hda-sdw-bpt helpers.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart committed:
soundwire: stream: extend sdw_alloc_stream() to take 'type' parameter

In the existing definition of sdw_stream_runtime, the 'type' member is
never set and defaults to PCM. To prepare for the BPT/BRA support, we
need to special-case streams and make use of the 'type'.

No functional change for now, the implicit PCM type is now explicit.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart committed:
soundwire: extend sdw_stream_type to BPT

BPT/BRA need to be special cased, i.e. there's no point in using the
bandwidth allocation since the entire frame can be used.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart committed:
soundwire: bus: add bpt_stream pointer

Add a convenience pointer to the 'sdw_bus' structure. BPT is a
dedicated stream which will typically not be handled by DAIs or
dailinks. Since there's only one BPT stream per link, storing the
pointer at the link level seems rather natural.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart committed:
soundwire: intel_auxdevice: add indirection for BPT send_async/wait

Mirror abstraction added for master ops.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart committed:
soundwire: stream: special-case the bus compute_params() routine

For BPT support, we want to allocate the entire audio payload and
bypass the allocation based on PCM/PDM parameters.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart committed:
ASoC: rt711-sdca: add DP0 support

DP0 is required for BPT/BRA transport.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart committed:
soundwire: stream: reuse existing code for BPT stream

DP0 (Data Port 0) is very similar to regular data ports, with minor
tweaks we can reuse the same code.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart committed:
soundwire: cadence: add BTP/BRA helpers to format data

The Cadence IP expects a specific format (detailed in the
Documentation). Add helpers to copy the data into the DMA buffer.

The crc8 table is for now only used by the Cadence driver. This table
might be moved to a common module at a later point if needed by other
controller implementations.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart committed:
soundwire: intel: add BPT context definition

This is needed to be shared between open/send_async/close.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart committed:
soundwire: cadence: add BTP support for DP0

The register definitions are missing a BULK_ENABLE bitfield which must
be set for DP0.

In addition, the existing mapping from PDI to Data Port is 1:1. That's
fine for PCM streams which are by construction in one direction
only. The BTP/BRA protocol is bidirectional and relies on DP0 only,
which breaks the 1:1 mapping. DP0 MUST be mapped to both PDI0 and
PDI1, with PDI0 taking care of the TX direction and PDI1 of the RX
direction.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao commented on drivers/soundwire/stream.c:
```diff
@@ -440,27 +456,41 @@ static int sdw_prep_deprep_slave_ports(struct sdw_bus *bus,
+		}
+
+		imp_def_interrupts = dpn_prop->imp_def_interrupts;
+		simple_ch_prep_sm = dpn_prop->simple_ch_prep_sm;
```
Need `ch_prep_timeout = dpn_prop->ch_prep_timeout;`?

bardliao commented on drivers/soundwire/stream.c:
```diff
@@ -2124,7 +2165,8 @@ int sdw_stream_add_slave(struct sdw_slave *slave,
 
-	ret = sdw_slave_port_config(slave, s_rt, port_config);
+	ret = sdw_slave_port_config(slave, s_rt, port_config,
+				    stream->type == SDW_STREAM_BPT ? true : false);
```
Simply to `stream->type == SDW_STREAM_BPT`?

bardliao commented on drivers/soundwire/stream.c:
```diff
@@ -2009,6 +2027,12 @@ int sdw_stream_add_master(struct sdw_bus *bus,
+			dev_err(bus->dev, "%s: Master runtime alloc failed for stream:%s: %d\n",
+				__func__, stream->name, ret);
+			goto unlock;
+		}
```
Should we merge the `if (IS_ERR(m_rt))` and the `if (!m_rt)` case?

bardliao commented on drivers/soundwire/cadence_master.c:
```diff
@@ -1982,5 +1983,905 @@ struct sdw_cdns_pdi *sdw_cdns_alloc_pdi(struct sdw_cdns *cdns,
+
+	total -= SDW_CDNS_BRA_DATA_CRC;
+
+	total -= SDW_CDNS_BRA_FOOTER_RESP;
```
Should we check if allocated_bytes_per_frame > these headers? I.e. total > 0

bardliao commented on drivers/soundwire/intel_ace2x.c:
```diff
@@ -13,12 +13,368 @@
+			msg->dev_num, msg->addr,
+			msg->len,
+			data_per_frame,
+			sdw->bpt_ctx.dmab_tx_bdl.area,
```
Should this be `sdw->bpt_ctx.dmab_rx_bdl.area`?

bardliao replied:
Ok, I see no matter read or write, this is to prepare the tx buffer.

bardliao commented on drivers/soundwire/intel_ace2x.c:
```diff
@@ -13,12 +13,368 @@
+			msg->len,
+			data_per_frame,
+			sdw->bpt_ctx.dmab_tx_bdl.area,
+			pdi0_buffer_size,
```
pdi1_buffer_size?

ranj063 commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,350 @@
+		dev_err(dev, "%s: hda_sdw_bpt_dma_prepare failed for RX: %d\n",
+			__func__, ret);
+
+		ret1 = hda_sdw_bpt_dma_deprepare(dev,
```
why do we need to do a deprepare if prepare failed in the first place?

bardliao replied:
We deprepare `bpt_tx_stream` while `bpt_rx_stream` prepare failed.

ranj063 commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,350 @@
+		dev_err(dev, "%s: hdac_bus_eml_sdw_map_stream_ch failed for TX: %d\n",
+			__func__, ret);
+
+	ret1 = hdac_bus_eml_sdw_map_stream_ch(sof_to_bus(sdev),
```
what is the reason for proceeding with this if ret < 0? We end up closing the BPT stream and returning the errror anyway further below isnt it?

bardliao replied:
Makes sense. Updated.

ranj063 commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,350 @@
+		dev_err(dev, "%s: hda_sdw_bpt_dma_enable failed for RX: %d\n",
+			__func__, ret);
+
+		ret1 = hda_sdw_bpt_dma_disable(dev, bpt_tx_stream);
```
again this seems unnecessary? if we never enabled it successfully why do we need to disable?

bardliao replied:
The (ret < 0) here is failed to enable `bpt_rx_stream`. But the `bpt_tx_stream` is enabled. That's why we need to disable `bpt_tx_stream` in this case.

ranj063 commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,350 @@
+							  false);
+		dev_err(dev, "%s: SDW BPT TX DMA did not complete: %ld\n",
+			__func__, tx_position);
+		ret = -ETIMEDOUT;
```
it looks like this timeout error will be ignored and overwritten anyway on line 315. Is that correct?

bardliao replied:
Nice catch, updated.

ranj063 commented on drivers/soundwire/intel.h:
```diff
@@ -48,11 +48,23 @@ struct sdw_intel_link_res {
 	struct hdac_bus *hbus;
 };
 
+struct sdw_intel_bpt {
```
do you want to add a description for this structure?

bardliao replied:
Yes.

ranj063 commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -64,6 +151,7 @@ static int hda_sdw_bpt_dma_prepare(struct device *dev,
 
 		snd_hdac_ext_bus_link_set_stream_id(hlink, stream_tag);
 	}
+
```
spurious change

bardliao replied:
fixed

ranj063 commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -104,8 +213,26 @@ static int hda_sdw_bpt_dma_enable(struct device *dev,
 	ret = hda_cl_trigger(sdev->dev, sdw_bpt_stream, SNDRV_PCM_TRIGGER_START);
 	if (ret < 0) {
 		dev_err(sdev->dev, "%s: SDW BPT DMA trigger start failed\n", __func__);
+
```
spurious change

bardliao replied:
fixed

ranj063 commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -115,11 +242,24 @@ static int hda_sdw_bpt_dma_disable(struct device *dev,
 		dev_err(sdev->dev, "%s: SDW BPT DMA trigger stop failed\n", __func__);
 		return ret;
 	}
+
```
same here

bardliao replied:
fixed

ranj063 commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -104,8 +213,26 @@ static int hda_sdw_bpt_dma_enable(struct device *dev,
+					hdac_stream(sdw_bpt_stream)->stream_tag,
+					hdac_stream(sdw_bpt_stream)->direction,
+					SOF_IPC4_PIPE_RUNNING);
+		if (ret < 0) {
```
dont you need to do hda_cl_trigger STOP in this error path?

bardliao replied:
> dont you need to do hda_cl_trigger STOP in this error path?

Yes, done.

plbossart commented on Documentation/driver-api/soundwire/summary.rst:
```diff
@@ -187,10 +187,7 @@ reconfigurations.
-
-
- (2) Multiple data lane support.
+ (1) Multiple data lane support.
```
didn't you provide multi-lane support already?

bardliao replied:
Indeed, that is great.

plbossart commented on drivers/soundwire/cadence_master.c:
```diff
@@ -1923,13 +1924,20 @@ void sdw_cdns_config_stream(struct sdw_cdns *cdns,
 		if (cdns->bus.params.m_data_mode != SDW_PORT_DATA_MODE_NORMAL)
 			val |= CDNS_PORTCTRL_TEST_FAILED;
+	} else if (pdi->num == 0) {
+		val |= CDNS_PORTCTRL_BULK_ENABLE;
```
double-check if this is correct, I wonder if pdi->num == 1 should also be dedicated to BULK.

bardliao replied:
Yes, changed to `if (pdi->num == 0 || pdi->num == 1)`

plbossart commented on drivers/soundwire/cadence_master.c:
```diff
@@ -1923,13 +1924,20 @@ void sdw_cdns_config_stream(struct sdw_cdns *cdns,
+	if (pdi->num == 1)
+		val = 0;
+	else
+		val = pdi->num;
```
see here PDI1 is used for bulk!

plbossart commented on drivers/soundwire/cadence_master.c:
```diff
@@ -1923,13 +1924,20 @@ void sdw_cdns_config_stream(struct sdw_cdns *cdns,
 		     val);
 
-	val = pdi->num;
+	/* The DataPort0 needs to be mapped to both PDI and PDI1 ! */
```
typo: PDI0 and PDI1.

bardliao replied:
fixed

plbossart commented on drivers/soundwire/cadence_master.c:
```diff
@@ -184,6 +184,7 @@ MODULE_PARM_DESC(cdns_mcp_int_mask, "Cadence MCP IntMask");
 #define CDNS_PORTCTRL_TEST_FAILED		BIT(1)
 #define CDNS_PORTCTRL_DIRN			BIT(7)
 #define CDNS_PORTCTRL_BANK_INVERT		BIT(8)
+#define CDNS_PORTCTRL_BULK_ENABLE		BIT(16)
```
typo in commit message:
PDI**0** taking care of the TX direction and PDI1 of the RX
direction.

bardliao replied:
fixed

plbossart commented on include/sound/hda-sdw-bpt.h:
```diff
@@ -0,0 +1,76 @@
+ * This file is provided under a dual BSD/GPLv2 license.  When using or
+ * redistributing this file, you may do so under either license.
+ *
+ * Copyright(c) 2023 Intel Corporation. All rights reserved.
```
update the year, it's now 2025

bardliao replied:
done

plbossart commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,352 @@
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2024 Intel Corporation. All rights reserved.
```
also remove "All rights reserved", not aligned with Intel recommendations.

bardliao replied:
done

plbossart commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,352 @@
+#include "../sof-priv.h"
+#include "hda.h"
+
+#define BPT_FREQUENCY		192000
```
need explanation of this value, I think it's 9.6MHz * Double data rate factor.
but if you can vary the clock now you'd need to use the max clock before using BRA/BPT...

bardliao replied:
@plbossart I think it is the sampling rate of the HDA stream. But I am not sure how does it related to the SoundWire bus rate. The SDW bus clock and the HDA sampling rate seems to be independent to me.
I guess why `192000` is because that is the max value of [rate_bits[]](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/hda/hdac_device.c#L706).

plbossart replied:
I can't recall where this came from... certainly something related to HDaudio format but why did we take 192k as the reference?

plbossart commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,352 @@
+							  false);
+		msleep(1);
+		i++;
+	} while (tx_position && i < HDA_BPT_IOC_TIMEOUT_MS);
```
probably need an error message if the DMA position was never cleared and the timeout happens?

bardliao replied:
done

plbossart commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,352 @@
+							  false);
+		msleep(1);
+		i++;
+	} while (rx_position && i < HDA_BPT_IOC_TIMEOUT_MS);
```
same comment on error message on timeout

bardliao replied:
done

ujfalusi replied:
you mean that we don't need error message here in case of timeout?

ujfalusi replied:
and why we are trying again if the first time we get timeout?

bardliao replied:
No, I added the error message in the wrong place, and I forgot to assign ret = -ETIMEDOUT;. Fixed it now.

plbossart commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,352 @@
+
+	ret1 = hda_sdw_bpt_dma_disable(dev, bpt_tx_stream);
+	if (!ret)
+		ret = ret1;
```
I can't remember if there's any reason to disable RX first?

bardliao replied:
I think it is because we enable TX first. So to be symmetrical, we disable RX first.

ranj063 commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,352 @@
+					     GENMASK(num_channels_rx - 1, 0),
+					     hdac_stream(*bpt_rx_stream)->stream_tag,
+					     SNDRV_PCM_STREAM_CAPTURE);
+	if (ret < 0)
```
I think we should return 0 if ret is 0 before this, so that you dont have to check again for ret < 0 on line 202

bardliao replied:
done

plbossart reviewed (commented):
nice update @bardliao, keep up the good work.

Don't forget to add your own Signed-off-by tag.

lgirdwood commented:
@ujfalusi can you give this a review too. Thanks !

ranj063 commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -341,12 +477,12 @@ int hda_sdw_bpt_close(struct device *dev,
 	ret = hda_sdw_bpt_dma_deprepare(dev,
-					bpt_tx_stream,
-					dmab_tx_bdl);
+					bpt_rx_stream,
```
@bardliao Im almost good with the series except for this change. Is this change related to the addition of chain DMA support or can this change be squashed into the previous patch where you introduce hda_sdw_bpt_close()?

bardliao replied:
Yes, it makes more sense to move to the commit that introduce hda_sdw_bpt_close(). Done.

ujfalusi commented on drivers/soundwire/stream.c:
```diff
@@ -1815,12 +1815,14 @@ static int set_stream(struct snd_pcm_substream *substream,
  * sdw_alloc_stream() - Allocate and return stream runtime
  *
  * @stream_name: SoundWire stream name
+ * @type: stream type (could be PCM, PDM or BPT)
```
SDW_STREAM_BPT is yet to be added, we only have definition for PCM and PDM

bardliao replied:
Swapped the commits order with the next commit.

ujfalusi commented on include/linux/soundwire/sdw.h:
```diff
@@ -150,12 +150,14 @@ enum sdw_dpn_pkg_mode {
 enum sdw_stream_type {
 	SDW_STREAM_PCM = 0,
 	SDW_STREAM_PDM = 1,
+	SDW_STREAM_BPT = 2,
```
probably swap the two patches or merge them?

bardliao replied:
swapped

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -610,8 +686,9 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+ * @bpt_stream: boolean to conditionally use dedicated bit allocation
  */
-int sdw_compute_params(struct sdw_bus *bus, struct sdw_stream_runtime *stream)
+int sdw_compute_params(struct sdw_bus *bus, struct sdw_stream_runtime *stream, bool bpt_stream)
```
why do we pass the `stream->type` as a separate parameter via `bpt_stream`?

bardliao replied:
Good suggestion. Removed.

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -620,6 +697,11 @@ int sdw_compute_params(struct sdw_bus *bus, struct sdw_stream_runtime *stream)
 	if (ret < 0)
 		return ret;
 
+	if (bpt_stream) {
```
```c
if (stream->type == SDW_STREAM_BPT) {
```
?

ujfalusi commented on drivers/soundwire/stream.c:
```diff
@@ -1340,6 +1377,11 @@ struct sdw_dpn_prop *sdw_get_slave_dpn_prop(struct sdw_slave *slave,
+	if (!port_num) {
+		dev_err(&slave->dev, "%s: port_num is zero\n", __func__);
+		return NULL;
+	}
```
isn't this is an unrelated fix?

bardliao replied:
We added port_num == 0 (DP0) support and treat DP0 differently as other port data ports in this commit. So, I think it is fair to add the change in this commit.

ujfalusi commented on drivers/soundwire/bus.c:
```diff
@@ -2038,3 +2038,32 @@ void sdw_clear_slave_status(struct sdw_bus *bus, u32 request)
+		       struct sdw_bpt_msg *msg)
+{
+	if (msg->len > SDW_BPT_MSG_MAX_BYTES)
+		return -EINVAL;
```
So, larger transfer are better to be handled as PIO single writes? ;)

bardliao replied:
If the codec supports PIO. :) Alternative way is to send more than 1 BPT streams.

ujfalusi replied:
in any case, I would print an error on why we refused the transfer.

bardliao replied:
Error trace added

ujfalusi commented on drivers/soundwire/bus.c:
```diff
@@ -2038,3 +2038,32 @@ void sdw_clear_slave_status(struct sdw_bus *bus, u32 request)
 }
 EXPORT_SYMBOL(sdw_clear_slave_status);
+
+int sdw_bpt_send_async(struct sdw_bus *bus,
```
This does not align with the documentation (sdw_bpt_message_send_async)

bardliao replied:
Changed it in the documentation.

ujfalusi commented on include/linux/soundwire/sdw.h:
```diff
@@ -1046,6 +1068,13 @@ int sdw_compare_devid(struct sdw_slave *slave, struct sdw_slave_id id);
+		       struct sdw_bpt_msg *msg);
+int sdw_bpt_wait(struct sdw_bus *bus,
+		 struct sdw_slave *slave,
+		 struct sdw_bpt_msg *msg);
```
It is not immediately clear if you need to pass the same `*msg` to both functions or in wait you pass different one which will hold the reply..

Would it be practical to add a `sdw_bpt_send_sync` wrapper to combine these two?

bardliao replied:
The msg are the same in sdw_bpt_send_async() and sdw_bpt_wait(). However, I think sdw_slave_fw_download() is more useful for codec drivers. I will do it after the PR is merged.

ujfalusi replied:
I would not hardcode a specific function to an API which is just to transfer x bytes to (and probably from, but that is unlikely 0 unless you can download some recording from the codec for analysis).

bardliao replied:
Hmm, I don't see any case that the 2 APIs will be used separately. I will merge the 2 APIs

ujfalusi commented on drivers/soundwire/cadence_master.c:
```diff
@@ -1982,5 +1983,913 @@ struct sdw_cdns_pdi *sdw_cdns_alloc_pdi(struct sdw_cdns *cdns,
+{
+	u8 crc;
+	int data_written;
+	u8 *last_byte;
```
nitpick: reorder?

bardliao replied:
updated

ujfalusi commented on drivers/soundwire/cadence_master.c:
```diff
@@ -1982,5 +1983,913 @@ struct sdw_cdns_pdi *sdw_cdns_alloc_pdi(struct sdw_cdns *cdns,
+	 * two bytes at a time did not seem to improve the performance
+	 */
+	int i;
+	int j;
```
nitpick: `int i, j;`

bardliao replied:
updated

ujfalusi commented on drivers/soundwire/cadence_master.c:
```diff
@@ -1982,5 +1983,913 @@ struct sdw_cdns_pdi *sdw_cdns_alloc_pdi(struct sdw_cdns *cdns,
+{
+	u8 crc;
+	int data_written;
+	u8 *last_byte;
```
nitpick: reorder?

bardliao replied:
updated

ujfalusi commented on drivers/soundwire/cadence_master.c:
```diff
@@ -1982,5 +1983,913 @@ struct sdw_cdns_pdi *sdw_cdns_alloc_pdi(struct sdw_cdns *cdns,
+	int dma_data_written;
+	int total_dma_data_written;
+	u8 counter;
+	int ret;
```
nitpick: reorder?

bardliao replied:
updated

ujfalusi commented on drivers/soundwire/cadence_master.c:
```diff
@@ -1982,5 +1983,913 @@ struct sdw_cdns_pdi *sdw_cdns_alloc_pdi(struct sdw_cdns *cdns,
+	header[0] |= (dev_num << 2);
+
+	p_data = &data[0];
+	p_dma_buffer = &dma_buffer[0];
```
```c
p_data = data;
p_dma_buffer = dma_buffer;
```

would work equally?

ujfalusi commented on drivers/soundwire/cadence_master.c:
```diff
@@ -1982,5 +1983,913 @@ struct sdw_cdns_pdi *sdw_cdns_alloc_pdi(struct sdw_cdns *cdns,
+	int dma_data_written;
+	int total_dma_data_written;
+	u8 counter;
+	int ret;
```
nitpick: reorder?

bardliao replied:
updated

ujfalusi commented on drivers/soundwire/cadence_master.c:
```diff
@@ -1982,5 +1983,913 @@ struct sdw_cdns_pdi *sdw_cdns_alloc_pdi(struct sdw_cdns *cdns,
+	header[0] |= GENMASK(7, 6);	/* header is active */
+	header[0] |= (dev_num << 2);
+
+	p_dma_buffer = &dma_buffer[0];
```
```
p_dma_buffer = dma_buffer;
```
?

nitpick^2:
And why not do this at declaration?

bardliao replied:
updated

ujfalusi commented on drivers/soundwire/cadence_master.c:
```diff
@@ -1982,5 +1983,913 @@ struct sdw_cdns_pdi *sdw_cdns_alloc_pdi(struct sdw_cdns *cdns,
+	int i;
+	u8 crc;
+	u8 b0;
+	u8 b1;
```
I still prefer to have these collapsed into a single line, but I know, the author prefers this style.

ujfalusi commented on include/sound/hda-sdw-bpt.h:
```diff
@@ -0,0 +1,76 @@
+					 struct hdac_ext_stream *bpt_tx_stream,
+					 struct hdac_ext_stream *bpt_rx_stream)
+{
+	return 0;
```
How would the caller know that nothing actually got transferred and it needs to try single writes/reads?

bardliao replied:
Changed the functions to
```c
{
        WARN_ONCE(1, "SoundWire BPT is disabled");
        return -ENOTSUP;
}
```
Not sure if `-ENOTSUP` is the best return value.

ujfalusi commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,359 @@
+		     struct hdac_ext_stream **bpt_rx_stream,
+		     struct snd_dma_buffer *dmab_rx_bdl,
+		     u32 bpt_rx_num_bytes,
+		     u32 rx_dma_bandwidth)
```
What is tx_dma_bandwidth / rx_dma_bandwidth? Is it a desired or something factual and how the caller gets this?

bardliao replied:
It is factual. They are calculated by the caller.

ujfalusi commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,359 @@
+				    dmab_bdl,
+				    false,
+				    direction,
+				    false);
```
you can condensate this a bit I guess?  (moving parameters into same line)

bardliao replied:
Yes, done.

ujfalusi commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,359 @@
+		dev_err(sdev->dev, "%s: SDW BPT DMA trigger start failed\n", __func__);
+		return ret;
+	}
+	return 0;
```
```c

	if (ret < 0)
		dev_err(sdev->dev, "%s: SDW BPT DMA trigger start failed\n", __func__);

	return ret;
```

bardliao replied:
done

ujfalusi commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,359 @@
+		dev_err(sdev->dev, "%s: SDW BPT DMA trigger stop failed\n", __func__);
+		return ret;
+	}
+	return 0;
```
```c

	if (ret < 0)
		dev_err(sdev->dev, "%s: SDW BPT DMA trigger stop failed\n", __func__);

	return ret;
```

ujfalusi commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,359 @@
+	snd_pcm_uframes_t rx_position;
+	int i;
+	int ret = 0;
+	int ret1;
```
nitpick: reorder?

bardliao replied:
done

ujfalusi commented on drivers/soundwire/intel.h:
```diff
@@ -48,11 +48,32 @@ struct sdw_intel_link_res {
+ * @bpt_rx_stream: BPT RX stream
+ * @dmab_rx_bdl: BPT RX buffer descriptor list
+ * @pdi0_buffer_size: PDI0 buffer size
+ * @pdi1_buffer_size: PDI1 buffer size
```
num_frames and data_per_frames does not need to be described?

bardliao replied:
Oh, yes. Fixed

ujfalusi commented on drivers/soundwire/intel_ace2x.c:
```diff
@@ -13,12 +13,368 @@
+	pconfig.ch_mask = BIT(0);
+
+	ret = sdw_stream_add_slave(slave, &sconfig,
+				   &pconfig, 1, stream);
```
can fit to one line

bardliao replied:
done

ujfalusi commented on drivers/soundwire/intel_ace2x.c:
```diff
@@ -13,12 +13,368 @@
+		return ret;
+	}
+
+	return 0;
```
return ret;

bardliao replied:
done

ujfalusi commented on drivers/soundwire/intel_ace2x.c:
```diff
@@ -13,12 +13,368 @@
+		goto bpt_close;
+	}
+
+	return 0;
```
```
if (ret == 0) or if (!ret)
	return 0;

dev_err(cdns->dev, "%s: sdw_prepare_%s_dma_buffer failed %d\n",
	__func__, command ? "read" : "write", ret);
```
?

bardliao replied:
yes, done

ujfalusi commented on drivers/soundwire/intel_ace2x.c:
```diff
@@ -13,12 +13,368 @@
+		return ret;
+	}
+
+	dev_dbg(cdns->dev, "%s: done\n", __func__);
```
Do we need to print this?

bardliao replied:
removed

ujfalusi commented on drivers/soundwire/intel_ace2x.c:
```diff
@@ -13,12 +13,368 @@
+			dev_err(cdns->dev, "%s: BPT Read failed %d\n", __func__, ret);
+	}
+
+	dev_dbg(cdns->dev, "BPT Transfer done\n");
```
I would drop this as well.

bardliao replied:
removed

ujfalusi commented on drivers/soundwire/debugfs.c:
```diff
@@ -161,6 +162,25 @@ static int set_command(void *data, u64 value)
+
+	dev_dbg(&slave->dev, "command type: %s\n", value ? "BRA" : "Column0");
+
+	cmd_type = value;
```
cmd_type is `int`, value is `u64`

ujfalusi commented on drivers/soundwire/debugfs.c:
```diff
@@ -196,9 +216,49 @@ static int set_num_bytes(void *data, u64 value)
+
+	ret = sdw_bpt_send_async(slave->bus,
+				 slave,
+				 &msg);
```
one line?

bardliao replied:
done

ujfalusi commented on drivers/soundwire/debugfs.c:
```diff
@@ -196,9 +216,49 @@ static int set_num_bytes(void *data, u64 value)
+	if (ret < 0)
+		dev_err(&slave->dev, "%s: bpt_wait failed: %d\n", __func__, ret);
+
+	dev_dbg(&slave->dev, "wait done\n");
```
I would check a log and drop debug prints to make it less noisy?

bardliao replied:
ok, removed

ujfalusi commented on sound/soc/codecs/rt711-sdca-sdw.c:
```diff
@@ -225,6 +225,14 @@ static int rt711_sdca_read_prop(struct sdw_slave *slave)
+		return -ENOMEM;
+
+	prop->dp0_prop->simple_ch_prep_sm = true;
+	prop->dp0_prop->ch_prep_timeout = 10;
```
No other codec can declare this? Only rt711-sdca?

bardliao replied:
We get the properties from sdw_slave_read_prop() for other codecs.

ujfalusi replied:
I mean in theory rt-715, 721, 722 could also?
What I meant is: why only rt711?

bardliao replied:
In theory, most SDW codecs support BPT. In fact, prop->dp0_prop->simple_ch_prep_sm should come from the "mipi-sdw-simplified-channel-prepare-sm" property. Only rt711 is because it is the first codec that we tested. I will get the supported codec list from Realtek and Cirrus and add the codecs to the PR.

ujfalusi commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -14,12 +14,82 @@
+		dma_id += ipc4_data->num_playback_streams;
+
+	msg.primary |=  SOF_IPC4_GLB_CHAIN_DMA_HOST_ID(dma_id);
+	msg.primary |=  SOF_IPC4_GLB_CHAIN_DMA_LINK_ID(dma_id);
```
This builds on the assumption that the host and link DMA is always allocated in matching pairs, but can we safely make this assumption?
Can we have two playback PCMs active towards a single DAI (two host and one link DMA)?

Or at this phase no audio stream can be active?

bardliao replied:
For BPT, they are always 1:1 mapping. And BPT stream and audio stream are mutually exclusive.

ujfalusi replied:
probably a comment would make this cleaner.

ujfalusi commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -115,6 +239,18 @@ static int hda_sdw_bpt_dma_disable(struct device *dev,
+			dev_err(sdev->dev, "%s: chain_dma_trigger PIPE_PAUSED failed: %d\n",
+				__func__, ret);
+	}
+
```
Should this patch to be moved earlier? can BRA work for rt711 before this patch?

bardliao replied:
> Should this patch to be moved earlier? can BRA work for rt711 before this patch?

Yes, rt711 can work without this in dsp less mode. But, yes, it is better to move the commit before the codec change.

ujfalusi replied:
it is unlikely that anyone would use DSPless mode apart from developers for testing.

ujfalusi reviewed (commented):
@bardliao, I think you should add your signed-off-by and if your changes tot he original code warrants it and if @plbossart agrees, probably a Co-developed as well?

bardliao committed:
ASoC: SOF: Intel: hda-sdw-bpt: add CHAIN_DMA support

When the firmware is involved, the data can be transferred with a
CHAIN_DMA on LNL+.

The CHAIN_DMA needs to be programmed before the DMAs per the
documentation. The states are not exactly symmetrical, on stop we must
do a PAUSE and RESET.

The FIFO size of 10ms was determined experimentally. With the minimum
of 2ms, errors were reported by the codec, likely because of xruns.

The code flow deals with the two TX and RX CHAIN_DMAs in symmetrical
ways, i.e.
alloc TX
alloc RX
enable TX
enable RX
disable RX
disable TX
free RX
free TX

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -87,6 +87,82 @@ void sdw_compute_slave_ports(struct sdw_master_runtime *m_rt,
+				     SDW_PORT_DATA_MODE_NORMAL);
+	}
+
+	sdw_compute_slave_dp0_ports(m_rt, &t_data);
```
it is unexpected to see that a slave is configured from a function which should be taking care of the master.

If you want to keep it like this then probably:
```
s/sdw_compute_dp0_port_params/sdw_compute_bpt_port_params
s/sdw_compute_dp0_master_ports/sdw_compute_dp0_ports_params
s/sdw_compute_slave_dp0_ports/sdw_compute_slave_dp0_ports_params
```
and add comment that in sdw_compute_dp0_ports_params() we take care of masters.
or
```c
static void sdw_compute_bpt_port_params(struct sdw_bus *bus)
{
	struct sdw_master_runtime *m_rt;

	list_for_each_entry(m_rt, &bus->m_rt_list, bus_node) {
		struct sdw_transport_data t_data = {0};

		sdw_compute_dp0_master_ports(m_rt, &t_data);
		sdw_compute_slave_dp0_ports(m_rt, &t_data);
	}
}
```

bardliao replied:
The naming is used in the existing code.
[sdw_compute_slave_ports()](https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/generic_bandwidth_allocation.c#L144) is called in [sdw_compute_master_ports()](https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/generic_bandwidth_allocation.c#L90)
But maybe we can merge sdw_compute_slave_dp0_ports() and sdw_compute_dp0_master_ports(). What do you think?

ujfalusi replied:
I was pointing out that you introduce a function `sdw_compute_dp0_master_ports()` which calls `sdw_compute_slave_dp0_ports()`

it is unexpected to see a master config function to do slave configuration and I just noticed that one is dp0_master the other is slave_dp0.

in other words, I think it is not intuitive and consistent.

ujfalusi replied:
You are not planning to address this?

bardliao replied:
fixed.

ujfalusi commented on drivers/soundwire/stream.c:
```diff
@@ -95,13 +95,18 @@ static int _sdw_program_slave_port_params(struct sdw_bus *bus,
 		return ret;
 	}
 
+skip_block_ctrl3:
```
I tend to prefer code w/o goto:
```c

	if (t_params->port_num) {
		/* Program DPN_BlockCtrl3 register */
		ret = sdw_write_no_pm(slave, addr2, t_params->blk_pkg_mode);
		if (ret < 0) {
			dev_err(bus->dev, "DPN_BlockCtrl3 register write failed\n");
			return ret;
		}
	}
```

bardliao replied:
fixed

ujfalusi commented on drivers/soundwire/bus.c:
```diff
@@ -2038,3 +2038,32 @@ void sdw_clear_slave_status(struct sdw_bus *bus, u32 request)
+	/* make sure all callbacks are defined */
+	if (!bus->ops->bpt_send_async ||
+	    !bus->ops->bpt_wait)
+		return -ENOTSUPP;
```
What would be the fallback strategy of a codec driver in case the platform they are integrated is not supporting BRA?

Should we have some sort of is_bpt_supported() exposed for clients so they can make the decision runtime?

bardliao replied:
Codec driver may use single SDW write or different IO interface like SPI. I would expect that codec driver test the return value of  sdw_bpt_send_async(). It will not take longer time to get -ENOTSUPP if the bus doesn't support BRA.

ujfalusi commented on drivers/soundwire/cadence_master.c:
```diff
@@ -1982,5 +1983,905 @@ struct sdw_cdns_pdi *sdw_cdns_alloc_pdi(struct sdw_cdns *cdns,
+	0x79, /* 4 */
+	0x34, /* 5 */
+	0xe3, /* 6 */
+	0xae, /* 7 */
```
if you group them to 8 to reduce the number of lines, I don't think anyone will check the values anyways?
```c
static const u8 sdw_crc8_lookup_msb[CRC8_TABLE_SIZE] = {
	0x00, 0x4d, 0x9a, 0xd7, 0x79, 0x34, 0xe3, 0xae, /* 0 - 7 */
	0xf2, 0xbf, 0x68, 0x25, 0x8b, 0xc6, 0x11, 0x5c, /* 8 - 15 */
	...
```
If it makes sense.

bardliao replied:
Yes, done

ujfalusi commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,352 @@
+	 * bandwidth requirements
+	 */
+	format |= (num_channels - 1);
+	format |= BPT_MULTIPLIER << AC_FMT_MULT_SHIFT;
```
```c
format |= HDA_RATE(48, 4, 1); /* 192KHz */
```
Instead of the magic definer?

bardliao replied:
But HDA_RATE is a local macro in sound/hda/hdac_device.c.

ujfalusi replied:
Oh, OK, I did not realized that

ujfalusi commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,352 @@
+	format |= (num_channels - 1);
+	format |= BPT_MULTIPLIER << AC_FMT_MULT_SHIFT;
+
+	dev_dbg(dev, "direction %d format_val %x\n", direction, format);
```
I would prefer `%#x` for the format

bardliao replied:
fixed

ujfalusi commented on drivers/soundwire/intel_ace2x.c:
```diff
@@ -13,12 +13,353 @@
+	struct sdw_cdns_pdi *pdi0;
+	struct sdw_cdns_pdi *pdi1;
+	const int ch = 1;
+	const int num_pdi = 2;
```
why not use the number instead of obfuscating?

ujfalusi replied:
ch and num_pdi is const, why do you need  a variable for them?

bardliao replied:
fixed

ujfalusi commented on drivers/soundwire/intel_ace2x.c:
```diff
@@ -13,12 +13,353 @@
+	stream = sdw_alloc_stream("BPT", SDW_STREAM_BPT);
+	if (!stream)
+		return -ENOMEM;
+	cdns->bus.bpt_stream = stream;
```
I would have a blank line between the return and this.

bardliao replied:
added

ujfalusi commented on drivers/soundwire/intel_ace2x.c:
```diff
@@ -13,12 +13,353 @@
+
+		intel_ace2x_bpt_close_stream(sdw, slave, msg);
+
+		return ret;
```
just use a single `return ret` at the end, sdw_enable_stream() returns 0 or negative on error.

bardliao replied:
done

ujfalusi commented on drivers/soundwire/intel_ace2x.c:
```diff
@@ -13,12 +13,353 @@
+		dev_err(cdns->dev, "%s: sdw_stream_enable failed: %d\n",
+			__func__, ret);
+
+		intel_ace2x_bpt_close_stream(sdw, slave, msg);
```
you don't need to clean that async out somehow if the enable fails?

bardliao replied:
Yeah, I think close the stream is enough.

ujfalusi commented on drivers/soundwire/debugfs.c:
```diff
@@ -189,16 +209,50 @@ static int set_num_bytes(void *data, u64 value)
+
+	ret = sdw_bpt_wait(slave->bus,
+			   slave,
+			   &msg);
```
ca fit to one line?

bardliao replied:
Yes, fixed

ujfalusi commented on drivers/soundwire/stream.c:
```diff
@@ -1027,9 +1062,14 @@ static int sdw_slave_port_config(struct sdw_slave *slave,
+			if (ret < 0)
+				return ret;
+		} else {
+			if (port_config[i].num)
```
this can be
```c
} else if (port_config[i].num) {
```

ujfalusi commented on drivers/soundwire/bus.c:
```diff
@@ -2038,3 +2038,38 @@ void sdw_clear_slave_status(struct sdw_bus *bus, u32 request)
+
+int sdw_bpt_wait(struct sdw_bus *bus,
+		 struct sdw_slave *slave,
+		 struct sdw_bpt_msg *msg)
```
nitpick: lots of things can be fit to use less lines, this for example can be in two lines.

bardliao replied:
fixed

ujfalusi commented on drivers/soundwire/stream.c:
```diff
@@ -1240,6 +1240,20 @@ static struct sdw_master_runtime
+	if (stream->type == SDW_STREAM_BPT) {
+		if (bus->stream_refcount > 0 || bus->bpt_stream_refcount > 0) {
+			dev_err(bus->dev, "%s: invalid configuration %d streams %d BPT streams\n",
+				__func__, bus->stream_refcount, bus->bpt_stream_refcount);
```
```c
			dev_err(bus->dev,
				"%s: invalid configuration %d streams %d BPT streams\n",
				__func__, bus->stream_refcount,
				bus->bpt_stream_refcount);
```
What are 'configuration streams' are?
perhaps
```c
			dev_err(bus->dev,
				"%s: %d/%d audio/BPT stream already allocated\n",
				__func__, bus->stream_refcount,
				bus->bpt_stream_refcount);
```
?

bardliao replied:
fixed

ujfalusi commented on include/linux/soundwire/sdw.h:
```diff
@@ -1020,6 +1021,7 @@ struct sdw_bus {
 	struct sdw_bus_params params;
 	int stream_refcount;
 	int bpt_stream_refcount;
+	struct sdw_stream_runtime *bpt_stream;
```
it is interesting that we don't need 'normal' stream pointer for convenience, only for bpt.

bardliao replied:
> it is interesting that we don't need 'normal' stream pointer for convenience, only for bpt.

We can get the stream pointer easily by snd_soc_dai_get_stream() for normal stream. But we don't involve ASoC in the BPT case.

ujfalusi replied:
so it is not for mere convenience, but necessity ;)

ujfalusi commented on include/linux/soundwire/sdw_intel.h:
```diff
@@ -436,6 +436,14 @@ struct sdw_intel_hw_ops {
+			      struct sdw_bpt_msg *msg);
+	int (*bpt_wait)(struct sdw_intel *sdw,
+			struct sdw_slave *slave,
+			struct sdw_bpt_msg *msg);
```
why not make these (in in other places to have few lines (these can fold to two/two lines)

bardliao replied:
fixed

ujfalusi commented on include/linux/soundwire/sdw_intel.h:
```diff
@@ -436,6 +436,14 @@ struct sdw_intel_hw_ops {
+			struct sdw_slave *slave,
+			struct sdw_bpt_msg *msg);
+
 };
```
no blank line

bardliao replied:
fixed

ujfalusi commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,352 @@
+				 *bpt_tx_stream,
+				 dmab_tx_bdl,
+				 *bpt_rx_stream,
+				 dmab_rx_bdl);
```
this looks a bit funny when all parameters are in new line...

bardliao replied:
fixed

ujfalusi commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,352 @@
+		return ret;
+	}
+
+	return 0;
```
I think you can drop the `return ret` from line 225 and use it here

bardliao replied:
fixed

ujfalusi commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -14,12 +14,82 @@
+ * with major simplifications since there are no pipelines defined
+ * and no dependency on ALSA hw_params
+ */
+static int chain_dma_trigger(struct snd_sof_dev *sdev,
```
I would move this patch even further forward, before the debugfs extension

bardliao replied:
fixed

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -87,6 +87,53 @@ void sdw_compute_slave_ports(struct sdw_master_runtime *m_rt,
+
+	list_for_each_entry(m_rt, &bus->m_rt_list, bus_node) {
+		sdw_compute_dp0_master_ports(m_rt);
+		sdw_compute_dp0_slave_ports(m_rt);
```
You had that `t_data` in previous versions which I cannot check anymore, that is no longer needed?

bardliao replied:
Right, the values of `t_data` are constant.

ujfalusi commented on drivers/soundwire/bus.c:
```diff
@@ -2038,3 +2038,37 @@ void sdw_clear_slave_status(struct sdw_bus *bus, u32 request)
 }
 EXPORT_SYMBOL(sdw_clear_slave_status);
+
+int sdw_bpt_send_async_and_wait(struct sdw_bus *bus, struct sdw_slave *slave,
```
iow, sdw_bpt_send_sync() ?

bardliao replied:
After reading the document in the first commit again. I would keep the send async and wait APIs separately.
```
One possible strategy to speed-up all initialization tasks would be to
start a BRA transfer for firmware download, then deal with all the
"regular" read/writes in parallel with the command channel, and last
to wait for the BRA transfers to complete. This would allow for a
degree of overlap instead of a purely sequential solution. As a
results, the BRA API must support async transfers and expose a
separate wait function.
```

ujfalusi replied:
Sure, but adding a helper to do a sync transfer does not mean that the async and wait cannot be used as described while helping users to avoid duplicated code.
The `do_bpt_sequence()` could be as simple as:
```c
static int do_bpt_sequence(struct sdw_slave *slave, bool write, u8 *buffer)
{
	struct sdw_bpt_msg msg = {0};

	msg.addr = start_addr;
	msg.len = num_bytes;
	msg.dev_num = slave->dev_num;
	if (write)
		msg.flags = SDW_MSG_FLAG_WRITE;
	else
		msg.flags = SDW_MSG_FLAG_READ;
	msg.buf = buffer;

	return sdw_bpt_send_sync(slave->bus, slave, &msg); /* Errors printed in the helper */
}
```

ujfalusi commented on include/linux/soundwire/sdw.h:
```diff
@@ -961,6 +980,8 @@ struct sdw_stream_runtime {
  * @params: Current bus parameters
  * @stream_refcount: number of streams currently using this bus
+ * @btp_stream_refcount: number of BTP streams currently using this bus (should
+ * be zero or one, multiple streams per link is not supported).
```
It can be a bool, right? bpt_stream_active,

bardliao replied:
It could be, but I prefer refcount, just like `stream_refcount`. BPT is still a stream.

ujfalusi commented on include/linux/soundwire/sdw.h:
```diff
@@ -981,6 +981,7 @@ struct sdw_stream_runtime {
  * @stream_refcount: number of streams currently using this bus
  * @btp_stream_refcount: number of BTP streams currently using this bus (should
+ * @bpt_stream: pointer stored for convenience.
  * be zero or one, multiple streams per link is not supported).
```
You cut the description for the previous member in half ;)
And it is not for convenience, the bpt_stream pointer is needed to be stored to handle BPT/BRA

bardliao replied:
fixed

ujfalusi commented on include/sound/hda-sdw-bpt.h:
```diff
@@ -0,0 +1,80 @@
+					 struct hdac_ext_stream *bpt_tx_stream,
+					 struct hdac_ext_stream *bpt_rx_stream)
+{
+	WARN_ONCE(1, "SoundWire BPT is disabled");
```
WARN will dump stack?

bardliao replied:
Yes, it will dump stack.

ujfalusi commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -0,0 +1,332 @@
+
+	ret1 = hda_sdw_bpt_dma_deprepare(dev,
+					 bpt_tx_stream,
+					 dmab_tx_bdl);
```
you still have these spread around, they can fit to one line.

ujfalusi commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -83,6 +172,25 @@ static int hda_sdw_bpt_dma_deprepare(struct device *dev, struct hdac_ext_stream
 
+	if (!sdev->dspless_mode_selected) {
+
+		/* Release CHAIN_DMA resources */
```
no need for blank line?

ujfalusi commented on sound/soc/sof/intel/hda-sdw-bpt.c:
```diff
@@ -46,6 +116,23 @@ static int hda_sdw_bpt_dma_prepare(struct device *dev, struct hdac_ext_stream **
+
+		snd_hdac_ext_stream_setup(bpt_stream, format);
+
+	}
```
no need for blank line

ujfalusi commented on drivers/soundwire/debugfs.c:
```diff
@@ -154,13 +155,32 @@ static int set_command(void *data, u64 value)
 
 	dev_dbg(&slave->dev, "command: %s\n", value ? "read" : "write");
-	cmd = value;
+	cmd = (int)value;
```
this is a sparate patch?

ujfalusi reviewed (commented):
@bardliao, not much to comment on the functionality and flow, but can you check the patches for lines that can be folded?

bardliao commented:
> @bardliao, not much to comment on the functionality and flow, but can you check the patches for lines that can be folded?

I checked and folded the lines. Hope I didn't miss any.

ujfalusi commented on include/sound/hda-sdw-bpt.h:
```diff
@@ -0,0 +1,64 @@
+				    struct hdac_ext_stream *bpt_rx_stream,
+				    struct snd_dma_buffer *dmab_rx_bdl)
+{
+	WARN_ONCE(1, "SoundWire BPT is disabled");
```
Probably a single WARN from hda_sdw_bpt_open() is enough to catch the attention?

bardliao replied:
Yes, but it is harmless to have WARN in every function.

ujfalusi commented on include/sound/hda-sdw-bpt.h:
```diff
@@ -0,0 +1,64 @@
+ * Copyright(c) 2025 Intel Corporation.
+ */
+
+#include <linux/device.h>
```
it is a good practice to protect the header with
```c
#ifndef __HDA_SDW_BPT_H
#define __HDA_SDW_BPT_H
...
#endif /* __HDA_SDW_BPT_H */
```

ujfalusi commented on drivers/soundwire/debugfs.c:
```diff
@@ -196,9 +216,40 @@ static int set_num_bytes(void *data, u64 value)
 			 set_num_bytes, "%llu\n");
 
+static int do_bpt_sequence(struct sdw_slave *slave, bool write,
+			   u8 *buffer)
```
this can fit to one line ;)

bardliao replied:
fixed

ujfalusi reviewed (commented):
@bardliao, I only really have few small comments left..

ujfalusi reviewed (approved):
@bardliao (and @plbossart ) thank you I think this looks pretty nice!

ranj063 commented:
SOFCI TEST

plbossart commented on Documentation/driver-api/soundwire/bra.rst:
```diff
@@ -0,0 +1,336 @@
+       that stream can connect master ports as well as all peripheral
+       DP0 ports.
+
+   (5) BRA transfers only make sense in the concept of one
```
in the context of

plbossart reviewed (approved):
one typo in comments, this can be handled with a fixup, thanks @bardliao

bardliao commented:
@plbossart The fixup PR: https://github.com/thesofproject/linux/pull/5305
